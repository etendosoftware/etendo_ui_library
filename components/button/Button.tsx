import React, {useState} from 'react';
import {Pressable, Text, View} from 'react-native';
import { NEUTRALS_100, PRIMARY_100, PRIMARY_80, QUATERNARY_100, QUATERNARY_50, SECONDARY_50 } from '../../styles/colors';
import {ButtonStyleVariant} from './Button.styles';
import {ButtonProps, ButtonStyleType} from './Button.types';

const Button = ({
  text,
  onPress,
  typeStyle,
  image,
  disabled,
  width = 'auto',
  height,
}: ButtonProps) => {
  const [isHovered, setIsHovered] = useState(false);

  const getHoveredBackgroundColor = (backgroundColor: ButtonStyleType) => {
    switch (backgroundColor) {
      case 'primary':
        return PRIMARY_80;
      case 'secondary':
        return SECONDARY_50;
      case 'terciary':
        return QUATERNARY_50;
      case 'whiteBorder':
        return PRIMARY_100;
    }
  };
  const getHoveredTextColorAndViewColor = (typeStyle: ButtonStyleType) => {
    switch (typeStyle) {
      case 'white':
        return QUATERNARY_100;
      case 'whiteBorder':
        return NEUTRALS_100;
      case 'primary':
        return NEUTRALS_100;
      default:
        return PRIMARY_100;
    }
  };

  const stateStyleContainer = () => {
    if (disabled) {
      return ButtonStyleVariant[typeStyle].containerDisabled;
    }
    if (isHovered) {
      const hoveredBackgroundColor = getHoveredBackgroundColor(typeStyle);
      return {
        ...ButtonStyleVariant[typeStyle].container,
        backgroundColor: hoveredBackgroundColor,
        width: width === 'full' ? '100%' : width,
      };
    }
    return ButtonStyleVariant[typeStyle].container;
  };
  const handleHoverIn = () => {
    setIsHovered(true);
  };

  const handleHoverOut = () => {
    setIsHovered(false);
  };

  const stateStyleText = () => {
    if (disabled) {
      return ButtonStyleVariant[typeStyle].textDisabled;
    }
    if (isHovered) {
      const hoveredTextColor = getHoveredTextColorAndViewColor(typeStyle);
      if (hoveredTextColor) {
        return {
          ...ButtonStyleVariant[typeStyle].text,
          color: hoveredTextColor,
        };
      }
    }
    return ButtonStyleVariant[typeStyle].text;
  };

  const newImage =
    image &&
    React.cloneElement(image, {
      fill: isHovered
        ? getHoveredTextColorAndViewColor(typeStyle)
        : ButtonStyleVariant[typeStyle].imageColor,
    });

  return (
    <Pressable
      onHoverIn={handleHoverIn}
      onHoverOut={handleHoverOut}
      style={[stateStyleContainer(), {height: height, width: width}]}
      onPress={onPress}
    >
      <View>{image && newImage}</View>
      <Text style={[stateStyleText()]}>{text}</Text>
    </Pressable>
  );
};

export default Button;
