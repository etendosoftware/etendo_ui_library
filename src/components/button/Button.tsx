import React, { useState } from 'react';
import {
  ActivityIndicator,
  Pressable,
  Text,
  TouchableOpacity,
} from 'react-native';
import {
  NEUTRAL_50,
  PRIMARY_100,
  SECONDARY_300,
  TERTIARY_50,
  TERTIARY_800,
} from '../../styles/colors';
import { ButtonProps, ButtonStyleType, IconStyleProps } from './Button.types';
import { ButtonStyleVariant, styles } from './Button.styles';

const Button = ({
  text,
  onPress,
  fontSize = 16,
  typeStyle,
  disabled,
  width = 'auto',
  height,
  iconLeft,
  iconRight,
  paddingHorizontal = 12,
  paddingVertical = 12,
  loading = false,
}: ButtonProps) => {
  const [isHovered, setIsHovered] = useState(false);

  const getHoveredBackgroundColor = (backgroundColor: ButtonStyleType) => {
    switch (backgroundColor) {
      case 'primary':
        return TERTIARY_800;
      case 'secondary':
        return SECONDARY_300;
      case 'terciary':
        return TERTIARY_50;
      case 'whiteBorder':
        return PRIMARY_100;
    }
  };
  const getHoveredTextColorAndViewColor = (typeStyle: ButtonStyleType) => {
    switch (typeStyle) {
      case 'white':
        return TERTIARY_800;
      case 'whiteBorder':
        return NEUTRAL_50;
      case 'primary':
        return NEUTRAL_50;
      default:
        return PRIMARY_100;
    }
  };

  const stateStyleContainer = (pressed: boolean) => {
    if (disabled) {
      return ButtonStyleVariant[typeStyle].containerDisabled;
    }
    if (pressed && !loading) {
      const hoveredBackgroundColor = getHoveredBackgroundColor(typeStyle);
      return {
        ...ButtonStyleVariant[typeStyle].container,
        opacity: 0.7,
        backgroundColor: hoveredBackgroundColor,
      };
    }
    if (isHovered && !loading) {
      const hoveredBackgroundColor = getHoveredBackgroundColor(typeStyle);
      return {
        ...ButtonStyleVariant[typeStyle].container,
        backgroundColor: hoveredBackgroundColor,
      };
    }
    return ButtonStyleVariant[typeStyle].container;
  };
  const handleHoverIn = () => {
    setIsHovered(true);
  };

  const handleHoverOut = () => {
    setIsHovered(false);
  };

  const stateStyleText = () => {
    if (loading && !iconLeft && !iconRight) {
      return { color: 'transparent' };
    }
    if (disabled) {
      return ButtonStyleVariant[typeStyle].textDisabled;
    }
    if (isHovered && !loading) {
      const hoveredTextColor = getHoveredTextColorAndViewColor(typeStyle);
      if (hoveredTextColor) {
        return {
          ...ButtonStyleVariant[typeStyle].text,
          color: hoveredTextColor,
        };
      }
    }
    return ButtonStyleVariant[typeStyle].text;
  };

  const stateStyleIcon = () => {
    if (disabled) {
      return ButtonStyleVariant[typeStyle].imageDisabled;
    }
    if (isHovered && !loading) {
      return isHovered
        ? getHoveredTextColorAndViewColor(typeStyle)
        : ButtonStyleVariant[typeStyle].imageColor;
    }
    return ButtonStyleVariant[typeStyle].imageColor;
  };

  const stateSizeIndicator = (iconComponent: React.ReactElement) => {
    if (iconComponent?.props?.style?.height) {
      return iconComponent?.props?.style?.height;
    }
    if (iconComponent?.props?.style?.width) {
      return iconComponent?.props?.style?.width;
    }
    return 16;
  };

  const renderIcon = (
    icon?: React.ReactElement,
    { marginRight = 0, marginLeft = 0 }: IconStyleProps = {},
  ) => {
    if (!icon) return null;

    const iconSize = stateSizeIndicator(icon);
    const iconColor = stateStyleIcon();

    if (loading) {
      return (
        <ActivityIndicator
          size={iconSize}
          color={iconColor}
          style={{ marginRight, marginLeft }}
        />
      );
    }

    return React.cloneElement(icon, {
      style: {
        width: icon?.props?.style?.width || 16,
        height: icon?.props?.style?.height || 16,
        marginRight,
        marginLeft,
      },
      fill: iconColor,
    });
  };

  const handleOnPress = () => {
    if (onPress) {
      onPress();
    }
  };
  return (
    <TouchableOpacity>
      <Pressable
        onHoverIn={handleHoverIn}
        onHoverOut={handleHoverOut}
        style={({ pressed }) => [
          stateStyleContainer(pressed),
          { height, width, paddingHorizontal, paddingVertical },
        ]}
        onPress={handleOnPress}>
        {renderIcon(iconLeft, {
          marginRight: iconLeft && text ? 8 : 0,
        })}
        {loading && !iconLeft && !iconRight && (
          <ActivityIndicator
            size={fontSize}
            color={stateStyleIcon()}
            style={styles.indicatorCenterStyle}
          />
        )}
        <Text style={[stateStyleText(), { fontSize }]}>{text}</Text>
        {renderIcon(iconRight, {
          marginLeft: iconRight && text ? 8 : 0,
        })}
      </Pressable>
    </TouchableOpacity>
  );
};

export default Button;
