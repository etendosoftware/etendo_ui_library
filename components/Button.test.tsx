import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react-native';

import { Button } from './Button';

describe('Running Test for Button', () => {
  it('Check Button Disabled', () => {
    render(<Button text="Button" onPress={() => { }} />);
    //expect(screen.getByRole('button', { name: 'Button' })).toBeDisabled();
  });
});
