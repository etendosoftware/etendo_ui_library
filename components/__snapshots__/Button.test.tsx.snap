// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Running Test for Button Check Button Disabled 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 1,
    }
  }
>
  <View
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "red",
        "borderRadius": 4,
        "flexGrow": 0,
        "opacity": 1,
        "paddingHorizontal": 16,
        "paddingVertical": 8,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontSize": 16,
            "fontWeight": "bold",
          },
          false,
        ]
      }
    >
      Button
    </Text>
  </View>
</View>
`;
